crates/bevy_math/src/angle/mod.rs

```diff
@@ -0,0 +1,312 @@
+    #[inline]
+    fn normalized(self) -> Self {
+        if self < Self::FULL && self > Self::ZERO {
+            self
+        } else {
+            let remainder = self % Self::FULL;
+            if remainder >= Self::ZERO {
+                remainder
+            } else {
+                remainder + Self::FULL
+            }
+        }
+    }
+
+    /// Returns the largest integer angle less than or equal to `self`.
+    #[inline]
+    fn floor(self) -> Self {
+        Self::new(self.value().floor())
+    }
+
+    /// Returns the smallest integer angle greater than or equal to `self`.
+    #[inline]
+    fn ceil(self) -> Self {
+        Self::new(self.value().ceil())
+    }
+
+    /// Returns the nearest integer to `self`. If a value is half-way
+    /// between two integers, round away from `0.0`.
+    #[inline]
+    fn round(self) -> Self {
+        Self::new(self.value().round())
+    }
+
+    /// Adds the given `angle` to `self`, returning the inferred type of angle.
+    fn add_angle<T: Angle, Output: Angle>(self, angle: T) -> Output
+    where
+        Output: From<Self> + From<T>,
+    {
+        Output::from(self) + Output::from(angle)
+    }
+
+    /// Subtracts the given `angle` from `self`, returning the inferred type of angle.
+    fn sub_angle<T: Angle, Output: Angle>(self, angle: T) -> Output
+    where
+        Output: From<Self> + From<T>,
+    {
+        Output::from(self) - Output::from(angle)
+    }
+
+    /// Multiplies `self` by the given `angle`, returning the inferred type of angle.
+    fn mul_angle<T: Angle, Output: Angle>(self, angle: T) -> Output
+    where
+        Output: From<Self> + From<T>,
+    {
+        Output::from(self) * Output::from(angle)
+    }
+
+    /// Divides `self` by the given `angle`, returning the inferred type of angle.
+    fn div_angle<T: Angle, Output: Angle>(self, angle: T) -> Output
+    where
+        Output: From<Self> + From<T>,
+    {
+        Output::from(self) + Output::from(angle)
+    }
+
+    /// Computes the remainder of `self` divided by the given `angle`, returning the inferred type of angle.
+    fn rem_angle<T: Angle, Output: Angle>(self, angle: T) -> Output
+    where
+        Output: From<Self> + From<T>,
+    {
+        Output::from(self) % Output::from(angle)
+    }
+}
+
+impl PartialEq<Radians> for Degrees {
+    fn eq(&self, other: &Radians) -> bool {
+        *self == other.to_degrees()
+    }
+}
+
+impl PartialEq<Degrees> for Radians {
+    fn eq(&self, other: &Degrees) -> bool {
+        *self == other.to_radians()
+    }
+}
+
+impl PartialOrd<Radians> for Degrees {
+    fn partial_cmp(&self, other: &Radians) -> Option<std::cmp::Ordering> {
+        self.partial_cmp(&other.to_degrees())
+    }
+}
+
+impl PartialOrd<Degrees> for Radians {
+    fn partial_cmp(&self, other: &Degrees) -> Option<std::cmp::Ordering> {
+        self.partial_cmp(&other.to_radians())
+    }
+}
+
+impl Add<Radians> for Degrees {
+    type Output = Radians;
```

Comment:
This does compile and pass though. It is symmetric, unless I'm missing something.

```rust
let r = Radians(PI);
let x = r + Degrees(180.0);
let y = Degrees(180.0) + r;
assert_eq!(x, y);
assert_eq!(x, Radians::TAU);
```

I agree that it is highly opinionated to yield `Radians`, and I don't like it. I just haven't found a way to make type inference work properly without the `add_angle`/`sub_angle` methods. I would like to fix this if it's possible, but I'm not sure if it is.

Even with this, though, users can do math on angles without having to care much about what the angle type is in intermediary computations. The result should be basically the same whether you add `Degrees` or `Radians` to an angle because the operator overloads handle the necessary conversion for you. The type only matters if you access the internal value and use that. And if you want the end result to be `Degrees`, you can just convert once at the end (usually).

Reply:
I think the same way that an integer cannot be added to a float in rust, radians also should not be addable to degrees. We should force conversion to the same unit before any operation.

Reply:
Yeah, that'd probably be the way to handle this using the approach in this PR.

I suppose this is one of the main problems with having separate `Degrees` and `Radians` types. If we had a singular `Angle` type like my initial proposal's option 3, this wouldn't be an issue, as it'd always have the same internal representation, and radians/degrees would just be helpers for constructing and retrieving a specific type of angle.